### PR TITLE
WIP ... Privacy Badger Refactor

### PR DESCRIPTION
**Pull request by cooperq**
removes dual privacy badger objects and logic around initiating a badger object any time you want to use it. Instead there is now a global badger and anything which would write to the disk first checks if the tab is incognito. 
Items I plan to finish tomorrow (9/5):
* add incognito mode test
* fix a couple remaining web extension bugs 

cc @pde @ghostwords @alexristich @rcutmore for review

### CODE REVIEW

File: src/background.js [hunk 1]
```diff
@@ -18,32 +18,29 @@
...
-var DEBUG = false;
-
 /**
 * privacy badger initializer
 */
-function Badger(tabData, isIncognito) {
-  this.isIncognito = isIncognito;
-  this.tabData = JSON.parse(JSON.stringify(tabData));
+function Badger() {
+  this.tabData = {};
```
Comment (ghostwords) [inline src/background.js]: Unnecessary?

File: src/options.js [hunk 2]
```diff
@@ -18,16 +18,14 @@
 
 var backgroundPage = chrome.extension.getBackgroundPage();
 var require = backgroundPage.require;
-var pb = backgroundPage.pb; // incognito_pb can't use this page
+var pb = backgroundPage.badger; // incognito_pb can't use this page
```
Comment (ghostwords) [inline src/options.js]: Remove comment about incognito_pb?

File: src/storage.js [hunk 3]
```diff
@@ -62,17 +64,16 @@ function BadgerPen(isIncognito, callback) {
...
     _.each(keys, function(key){
       if(store.hasOwnProperty(key)){
-        bp[key] = new BadgerStorage(isIncognito, key, store[key]);
+        bp[key] = new BadgerStorage(key, store[key]);
       } else {
-        var storage_obj = new BadgerStorage(isIncognito, key, {});
+        var storage_obj = new BadgerStorage(key, {});
         bp[key] = storage_obj;
-        _syncStorage(storage_obj);
+        _syncStorage(storage_obj, 0);
```
Comment (ghostwords) [inline src/storage.js]: Second arg is typo?

File: src/webrequest.js [hunk 4]
```diff
@@ -313,7 +289,9 @@ function recordSuperCookie(sender, msg) {
   var frameHost = window.extractHostFromURL(msg.docUrl); // docUrl: url of the frame with supercookie
   var frameOrigin = window.getBaseDomain(frameHost);
   var pageHost = window.extractHostFromURL(getFrameUrl(sender.tab.id, 0));
-  var badger = getBadgerWithTab(sender.tab.id);
+  if (incognito.tabIsIncognito(sender.tab.id)) {
```
Comment (ghostwords) [inline src/webrequest.js]: Move to top of function? Or, replace by checking https://developer.chrome.com/extensions/extension#property-inIncognitoContext inside content script and bailing early if incognito? The idea is to quit as soon as possible. Ideally would refrain from injecting content script into incognito tabs altogether ...

File: src/webrequest.js [hunk 5]
```diff
@@ -351,6 +329,9 @@ function recordFingerprinting(tabId, msg) {
   if (!isThirdPartyDomain(script_host, document_host)) {
     return;
   }
+  if (incognito.tabIsIncognito(tabId)){
```
Comment (ghostwords) [inline src/webrequest.js]: See comment for supercookie function above.

File: src/heuristicblocking.js [hunk 6]
```diff
@@ -527,7 +523,6 @@ function startListeners() {
     }
     if(hasSetCookie) {
       //var origin = window.getBaseDomain(Utils.makeURI(details.url).host);
```
Comment (rcutmore) [inline src/heuristicblocking.js]: Remove commented code?

File: src/background.js (hunk 1 above)
Comment (cooperq) [inline src/background.js]: fixed.

File: src/options.js (hunk 2 above)
Comment (cooperq) [inline src/options.js]: fixed

File: src/storage.js (hunk 3 above)
Comment (cooperq) [inline src/storage.js]: fixed

File: src/webrequest.js (hunk 4 above)
Comment (cooperq) [inline src/webrequest.js]: fixed

File: src/webrequest.js (hunk 5 above)
Comment (cooperq) [inline src/webrequest.js]: fixed

File: src/popup.js [hunk 7]
```diff
@@ -21,48 +21,48 @@
...
-  getTab(function(tab) {
-    var badger = backgroundPage.getBadgerWithTab(tab.id);
-    var settings = badger.storage.getBadgerStorageObject('settings_map');
-    var seenComic = settings.getItem("seenComic") || false;
-
-    function _setSeenComic() {
-      settings.setItem("seenComic", "true");
-    }
+  function _setSeenComic() {
+    settings.setItem("seenComic", "true");
```
Comment (cooperq) [inline src/popup.js]: Makes sense to me. Fixed.

**Comment by alexristich** [comment]
With suggested fixes, refactoring looks good to me. Also, I didn't realize how many times getBadgerWithTab was called. Looks way cleaner now 👍

File: src/options.js (hunk 2 above)
Comment (cooperq) [inline src/options.js]: I thought that too, but I was feeling lazy :smile: 

fixed.

**Comment by ghostwords** [comment]
Looks good! :shipit:

**Comment by rcutmore** [comment]
Looks good to me too

**Pull request merged**